File: deploy/cloudflare-worker/src/services/zscore.ts
export interface ZScoreMetricStats {
  value: number | null;
  mean: number | null;
  stddev: number | null;
  zscore: number | null;
  percentile: number | null;
  window_n: number;
  window_start: string | null;
  window_end: string | null;
}

export type ZScoreBand = 'cheap' | 'normal' | 'expensive' | 'na';

function round(value: number, precision = 4): number {
  const scale = Math.pow(10, precision);
  return Math.round(value * scale) / scale;
}

export function zscoreBand(z: number | null): ZScoreBand {
  if (z == null || Number.isNaN(z)) return 'na';
  if (z <= -0.5) return 'cheap';
  if (z >= 0.5) return 'expensive';
  return 'normal';
}

export function computeZScoreStats(
  currentValue: number | null,
  windowValues: number[],
  years: string[] = [],
): ZScoreMetricStats {
  const cleanWindow = windowValues.filter((value) => Number.isFinite(value));
  if (!cleanWindow.length) {
    return {
      value: currentValue,
      mean: null,
      stddev: null,
      zscore: null,
      percentile: null,
      window_n: 0,
      window_start: years.length ? years[0] : null,
      window_end: years.length ? years[years.length - 1] : null,
    };
  }

  const n = cleanWindow.length;
  const mean = cleanWindow.reduce((sum, value) => sum + value, 0) / n;
  const variance = cleanWindow.reduce((sum, value) => sum + Math.pow(value - mean, 2), 0) / n;
  const stddev = Math.sqrt(variance);

  let zscore: number | null = null;
  let percentile: number | null = null;
  if (currentValue != null && Number.isFinite(currentValue)) {
    zscore = stddev > 0 ? (currentValue - mean) / stddev : 0;
    const sorted = [...cleanWindow].sort((a, b) => a - b);
    let belowOrEqual = 0;
    for (const value of sorted) {
      if (value <= currentValue) belowOrEqual += 1;
    }
    percentile = (belowOrEqual / sorted.length) * 100;
  }

  return {
    value: currentValue,
    mean: round(mean),
    stddev: round(stddev),
    zscore: zscore == null ? null : round(zscore),
    percentile: percentile == null ? null : round(percentile, 2),
    window_n: n,
    window_start: years.length ? years[0] : null,
    window_end: years.length ? years[years.length - 1] : null,
  };
}

File: backend/app/services/zscore.py
"""Z-score helpers for historical valuation context."""
from __future__ import annotations

import math
from typing import Iterable, Sequence


def zscore_band(z_value: float | None) -> str:
    if z_value is None or math.isnan(z_value):
        return "na"
    if z_value <= -0.5:
        return "cheap"
    if z_value >= 0.5:
        return "expensive"
    return "normal"


def _round(value: float, precision: int = 4) -> float:
    return round(value, precision)


def compute_zscore_stats(
    current_value: float | None,
    window_values: Iterable[float],
    years: Sequence[str] | None = None,
) -> dict:
    clean = [float(v) for v in window_values if v is not None and math.isfinite(float(v))]
    years = list(years or [])

    if not clean:
        return {
            "value": current_value,
            "mean": None,
            "stddev": None,
            "zscore": None,
            "percentile": None,
            "window_n": 0,
            "window_start": years[0] if years else None,
            "window_end": years[-1] if years else None,
            "band": "na",
        }

    n = len(clean)
    mean = sum(clean) / n
    variance = sum((value - mean) ** 2 for value in clean) / n
    stddev = math.sqrt(variance)

    zscore = None
    percentile = None
    if current_value is not None and math.isfinite(float(current_value)):
        if stddev > 0:
            zscore = (float(current_value) - mean) / stddev
        else:
            zscore = 0.0
        sorted_values = sorted(clean)
        below_or_equal = sum(1 for value in sorted_values if value <= float(current_value))
        percentile = (below_or_equal / len(sorted_values)) * 100.0

    return {
        "value": current_value,
        "mean": _round(mean),
        "stddev": _round(stddev),
        "zscore": _round(zscore) if zscore is not None else None,
        "percentile": _round(percentile, 2) if percentile is not None else None,
        "window_n": n,
        "window_start": years[0] if years else None,
        "window_end": years[-1] if years else None,
        "band": zscore_band(zscore),
    }
